core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java

```diff
@@ -0,0 +1,402 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+
+import alluxio.ConfigurationRule;
+import alluxio.conf.Configuration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.status.CancelledException;
+import alluxio.exception.status.DeadlineExceededException;
+import alluxio.exception.status.InternalException;
+import alluxio.grpc.BlockMasterWorkerServiceGrpc;
+import alluxio.grpc.ConfigProperty;
+import alluxio.grpc.GrpcChannel;
+import alluxio.grpc.GrpcChannelBuilder;
+import alluxio.grpc.GrpcServer;
+import alluxio.grpc.GrpcServerAddress;
+import alluxio.grpc.GrpcServerBuilder;
+import alluxio.grpc.GrpcService;
+import alluxio.grpc.RegisterWorkerPRequest;
+import alluxio.grpc.RegisterWorkerPResponse;
+import alluxio.grpc.Scope;
+import alluxio.grpc.ServiceType;
+import alluxio.security.authentication.AuthType;
+import alluxio.security.user.ServerUserState;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Channel;
+import io.grpc.stub.StreamObserver;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public class RegisterStreamerTest {
+  private static final InetSocketAddress TEST_ADDRESS = new InetSocketAddress("localhost", 9999);
+
+  @Rule
+  public ConfigurationRule mConfigurationRule =
+      new ConfigurationRule(
+          new ImmutableMap.Builder<PropertyKey, Object>()
+          // set response timeout short to ensure that tests don't take too long
+          .put(PropertyKey.WORKER_REGISTER_STREAM_RESPONSE_TIMEOUT, 500)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_DEADLINE, 2000)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_COMPLETE_TIMEOUT, 500)
+          // set batch size to 1 so that the stream will send a request
+          // for every block, used for easily control the concurrency of the stream
+          .put(PropertyKey.WORKER_REGISTER_STREAM_BATCH_SIZE, 1)
+          // we don't handle authentication in testing
+          .put(PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL)
+          .build(),
+          Configuration.modifiableGlobal());
+
+  private GrpcServer mServer;
+  private GrpcChannel mChannel;
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutConcurrentRequest() throws Exception {
+    // create a server that does not respond to requests
+    createMockService(
+        new BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase() {
+          @Override
+          public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
+              StreamObserver<RegisterWorkerPResponse> responseObserver) {
```

Comment:
Can create a `WorkerRegistrationServiceHandler` class that is subclass of `BlockMasterWorkerServiceImplBase` and takes 3 handler methods that are `onNext` etc., and make `createMockService` take an instance of this class.

Reply:
I didn't quite understand the suggestion.. Are you suggesting creating a class like the following?
```
class WorkerRegistrationServiceHandler extends BlockMasterWorkerServiceImplBase {
  WorkerRegistrationServiceHandler(function onNext, function onError, function onCompleted) {
  }
}
```

Reply:
I thought about this again, there might be some danger in the encapsulation we're trying to make.
The below class:
```java
class WorkerRegistrationServiceHandler 
    extends BlockMasterWorkerServiceImplBase {

    StreamObserver mObserver;

    WorkerRegistrationServiceHandler(StreamObserver observer) {
        mObserver = observer;
    }

    @Override
      public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
          StreamObserver<RegisterWorkerPResponse> responseObserver) {
        //! BIG DANGER: mObserver will re re-used for a second call
        return mObserver;
      }
}
```
is NOT correct in the strict sense of a GRPC handler, as it reused possibly closed streams for further requests. The test functions in fact make one requests, but this code could easily be a problem if someone re-used it for new test functions.

The most decent solution would be to take, not an observer instance, but a function to create an instance. But this requires something like Factory pattern in java(:<), which could really result in more boilerplate than we're trying to eliminate. 

For now I would probably suggest staying with the boilerplate with the improvement of `NoOpStreamObserver`

Reply:
An alternative is to make a check inside `registerWorkerStream` method to only handle the first request. But this is still awkward and can easily create problems for new tests.

Reply:
Yeah that's a good idea, but `StreamObserver` need to access a `responseObserver` to propogate response to the other side. There's still the problem how this observer can be passed into the `NoOpStreamObserver`.
```java
private static class TestRegistrationServiceHandler
      extends BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase {
    private Supplier<StreamObserver<RegisterWorkerPRequest>> mSupplier;

    private TestRegistrationServiceHandler(
        Supplier<StreamObserver<RegisterWorkerPRequest>> supplier) {
      mSupplier = supplier;
    }

    @Override
    public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
        StreamObserver<RegisterWorkerPResponse> responseObserver) {
      // how to pass the responseObserver to the stream?
      return mSupplier.get();
    }
  }
```

Reply:
I believe Bowen's suggestion (and the consensus from our engineers) is to NOT use mocks. The reason is, if you have a class Foo, and you `mock(Foo.class)`, then later Foo implementation is changed, the change will be silently mocked. Then in a test against mocked Foo, our assumptions may not hold true anymore. An extended child class, on contrast will complain if there's an API change, so your tests fail and you know where to look at.

@YangchenYe323 On your concern, generally we don't worry about that too much. A child class used merely as a mock does not need to be logically complete. If the implementation makes it hard to be reused, the easiest way out is to add a comment saying do not reuse in other tests because of blah.

Reply:
I've made a fix in the new commit. Actually mocking is redundant in this test class, because we have been creating a real `BlockMasterWorkerServiceImplBase` anyway.